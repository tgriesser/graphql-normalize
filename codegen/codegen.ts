import type { CodegenPlugin } from '@graphql-codegen/plugin-helpers';

const NormalizeCodegenPlugin: CodegenPlugin = {
  plugin(schema, documents, config, info) {
    //
  },
};

export default NormalizeCodegenPlugin;
