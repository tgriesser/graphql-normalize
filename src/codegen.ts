import { NormalizeCodegenPlugin } from './codegen/plugin.js';
export type { TypePolicies } from './codegen/getCacheKey.js';
export type { NormalizeMetaShape, NormalizedDocShape } from './metadataShapes.js';

export { NormalizeCodegenPlugin } from './codegen/plugin.js';
export { generateNormalizedMetadataForDocs, generateNormalizedMetadata } from './codegen/generateNormalizedMetadata.js';
export { generateNormalizedOperations, generateNormalizedOperation } from './codegen/generateNormalizedOperation.js';

export { NormalizeCodegenPlugin as plugin };

export default {
  plugin: NormalizeCodegenPlugin,
};
