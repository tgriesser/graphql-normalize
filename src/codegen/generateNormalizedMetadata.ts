import { optimizeDocuments } from '@graphql-tools/relay-operation-optimizer';
import {
  DocumentNode,
  GraphQLSchema,
  Kind,
  TypeInfo,
  isListType,
  isWrappingType,
  visit,
  print,
  visitWithTypeInfo,
  ArgumentNode,
  FieldNode,
  OperationTypeNode,
  isObjectType,
  isScalarType,
  InlineFragmentNode,
  getNamedType,
  assertObjectType,
} from 'graphql';
import type { FieldDef, FieldMeta, NormalizeMetaShape, UnionMeta, VariableMeta } from '../metadataShapes';
import { TypePolicies, getCacheKey } from '../typePolicies';
import { stringifyVariables } from '../stringifyVariables';

/**
 * Given an operation and a schema, generates the metadata necessary to
 * write the operation into the cache
 */
export function generateNormalizedMetadata(
  schema: GraphQLSchema,
  operation: DocumentNode,
  typePolicies: TypePolicies = {}
): NormalizeMetaShape {
  let finalOperation = operation;
  if (operation.definitions.some((d) => d.kind === Kind.FRAGMENT_DEFINITION)) {
    finalOperation = optimizeDocuments(schema, [operation], {
      noLocation: true,
    })[0];
  }

  const typeInfo = new TypeInfo(schema);
  const normalizedDoc: NormalizeMetaShape = {
    operation: OperationTypeNode.QUERY,
    variables: [],
    fields: [],
  };

  let parentStack: Array<FieldMeta | UnionMeta | NormalizeMetaShape> = [];
  let parentFieldDef: FieldMeta | NormalizeMetaShape | UnionMeta = normalizedDoc;

  function popStack() {
    const parent = parentStack.pop();
    if (parent) {
      parentFieldDef = parent;
    }
  }

  function pushField(field: string | FieldDef) {
    parentFieldDef.fields ??= [];
    parentFieldDef.fields.push(field);
    if (typeof field !== 'string') {
      parentStack.push(parentFieldDef);
      parentFieldDef = field;
    }
  }

  function normalizeArgs(nodes: readonly ArgumentNode[]) {
    let hasVar = false;
    const argsDef: Record<string, any> = {};
    for (const arg of nodes) {
      if (arg.value.kind === Kind.VARIABLE) {
        argsDef[`$${arg.value.name.value}`] = arg.name.value;
        hasVar = true;
      } else {
        argsDef[arg.name.value] = print(arg.value);
      }
    }
    return hasVar ? argsDef : stringifyVariables(argsDef);
  }

  const visitor = visitWithTypeInfo(typeInfo, {
    OperationDefinition(op) {
      normalizedDoc.operation = op.operation;
    },
    VariableDefinition(node) {
      const obj: VariableMeta = {
        name: node.variable.name.value,
      };
      if (node.defaultValue) {
        obj.defaultValue = print(node.defaultValue);
      }
      normalizedDoc.variables.push(obj);
    },
    Field: {
      enter(node) {
        const { gqlType, listDepth } = unpackType(typeInfo);

        // If this is a boring ol scalar, we push it as a string
        if (isBoringScalar(node, typeInfo)) {
          pushField(node.name.value);
          return;
        }

        const fieldMeta: FieldMeta = { name: node.name.value };
        pushField(fieldMeta);

        if (listDepth) fieldMeta.list = listDepth;
        if (node.alias?.value) fieldMeta.alias = node.alias.value;

        if (isObjectType(gqlType)) {
          const cacheKey = getCacheKey(typePolicies, gqlType);
          if (cacheKey) {
            fieldMeta.cacheKey = cacheKey;
          }
        }
        if (node.arguments?.length) {
          fieldMeta.args = normalizeArgs(node.arguments);
        }

        const skipDirective = node.directives?.find((d) => d.name.value === 'skip');
        const includeDirective = node.directives?.find((d) => d.name.value === 'include');
        if (skipDirective) {
          fieldMeta.skip = normalizeArgs(skipDirective.arguments ?? []);
        }
        if (includeDirective) {
          fieldMeta.include = normalizeArgs(includeDirective.arguments ?? []);
        }
      },
      leave(node) {
        // Pop things back into place
        if (!isBoringScalar(node, typeInfo)) {
          popStack();
        }
      },
    },
    InlineFragment: {
      enter(node) {
        if (
          node.typeCondition?.name.value &&
          isConcreteAbstract(node, typeInfo, schema) &&
          parentIsFieldMeta(parentFieldDef)
        ) {
          const unionType: UnionMeta = {
            fields: [],
          };
          const outputType = getNamedType(typeInfo.getType());
          const cacheKey = getCacheKey(typePolicies, assertObjectType(outputType));
          if (cacheKey) {
            unionType.cacheKey = cacheKey;
          }

          parentFieldDef.possible ??= {};
          parentFieldDef.possible[node.typeCondition?.name.value] = unionType;
          parentStack.push(parentFieldDef);
          parentFieldDef = unionType;
        }
      },
      leave(node) {
        if (node.typeCondition?.name.value && isConcreteAbstract(node, typeInfo, schema)) {
          popStack();
        }
      },
    },
    FragmentDefinition() {
      throw noFragments();
    },
  });

  visit(finalOperation, visitor);

  return normalizedDoc;
}

function unpackType(typeInfo: TypeInfo) {
  let listDepth = 0;
  // We want to unwrap any list types, and make those the containing
  let gqlType = typeInfo.getType();
  while (isWrappingType(gqlType)) {
    if (isListType(gqlType)) {
      ++listDepth;
    }
    gqlType = gqlType.ofType;
  }
  return { gqlType, listDepth };
}

function isBoringScalar(node: FieldNode, typeInfo: TypeInfo) {
  const { gqlType, listDepth } = unpackType(typeInfo);
  return (
    isScalarType(gqlType) &&
    !listDepth &&
    !node.alias &&
    !node.arguments?.length &&
    !node.directives?.some((d) => d.name.value === 'skip' || d.name.value === 'includes')
  );
}

function noFragments() {
  return new Error('Fragment definitions have been stripped');
}

function isConcreteAbstract(node: InlineFragmentNode, typeInfo: TypeInfo, schema: GraphQLSchema) {
  if (node.typeCondition?.name) {
    const parentTypeName = typeInfo.getParentType()?.name;
    if (node.typeCondition.name.value !== parentTypeName) {
      return isObjectType(schema.getType(node.typeCondition?.name.value));
    }
  }
  return false;
}

function parentIsFieldMeta(parent: FieldMeta | NormalizeMetaShape | UnionMeta): parent is FieldMeta {
  return 'name' in parent;
}
