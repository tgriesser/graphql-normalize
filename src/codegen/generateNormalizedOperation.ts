import { optimizeDocuments } from '@graphql-tools/relay-operation-optimizer';
import { DocumentNode, GraphQLSchema, Kind } from 'graphql';
import { addNormalizingFields } from './addNormalizingFields';

/**
 * Given a schema & operation, flattens the fragments and adds any necessary missing fields
 * in order to properly normalize the response in a cache.
 */
export function generateNormalizedOperation(schema: GraphQLSchema, operation: DocumentNode) {
  let finalOperation = operation;
  if (operation.definitions.some((d) => d.kind === Kind.FRAGMENT_DEFINITION)) {
    finalOperation = optimizeDocuments(schema, [operation], {
      noLocation: true,
    })[0];
  }
  return addNormalizingFields(schema, finalOperation);
}
