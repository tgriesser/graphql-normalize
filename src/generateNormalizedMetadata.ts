import { optimizeDocuments } from '@graphql-tools/relay-operation-optimizer';
import {
  DocumentNode,
  GraphQLSchema,
  Kind,
  TypeInfo,
  isAbstractType,
  isListType,
  isWrappingType,
  visit,
  print,
  visitWithTypeInfo,
  ArgumentNode,
  FieldNode,
  SelectionSetNode,
  OperationTypeNode,
  isObjectType,
} from 'graphql';
import stableStringify from 'fast-json-stable-stringify';
import type { FieldDef, FieldMeta, NormalizeMeta, NormalizedDoc, VariableMeta } from './extensionShape';
import { TypePolicies, getCacheKey } from './typePolicies';

/**
 * Given an operation and a schema, generates the metadata necessary to
 * write the operation into the cache
 */
export function generateNormalizedMetadata(
  schema: GraphQLSchema,
  operation: DocumentNode,
  typePolicies: TypePolicies = {}
): NormalizedDoc {
  let finalOperation = operation;
  if (operation.definitions.some((d) => d.kind === Kind.FRAGMENT_DEFINITION)) {
    finalOperation = optimizeDocuments(schema, [operation], {
      noLocation: true,
    })[0];
  }

  const typeInfo = new TypeInfo(schema);
  const normalizedDoc: NormalizedDoc = {
    operation: OperationTypeNode.QUERY,
    variables: [],
    selectionSet: {},
  };

  const fieldPath: any[] = [];

  function setPath<K extends keyof NormalizeMeta>(key: K, value: NormalizeMeta[K]) {
    normalizedDoc.selectionSet[fieldPath.join('.')] ??= {};
    normalizedDoc.selectionSet[fieldPath.join('.')][key] = value;
  }
  function normalizeArgs(nodes: readonly ArgumentNode[]) {
    let hasVar = false;
    const argsDef: Record<string, any> = {};
    for (const arg of nodes) {
      if (arg.value.kind === Kind.VARIABLE) {
        argsDef[`$${arg.name.value}`] = arg.name.value;
        hasVar = true;
      } else {
        argsDef[arg.name.value] = print(arg.value);
      }
    }
    return hasVar ? argsDef : stableStringify(argsDef);
  }
  function normalizeField(node: FieldNode) {
    if (node.alias || node.arguments?.length) {
      const field: FieldMeta = {
        name: node.name.value,
      };
      if (node.alias) field.alias = node.alias.value;
      if (node.arguments?.length) field.args = normalizeArgs(node.arguments);
      return field;
    }
    return node.name.value;
  }
  function addSelectionFields(node: SelectionSetNode) {
    const fields: FieldDef[] = [];
    for (const sel of node.selections) {
      if (sel.kind === Kind.FRAGMENT_SPREAD) {
        throw noFragments();
      }
      if (sel.kind === Kind.FIELD) {
        fields.push(normalizeField(sel));
      }
      // If we have an inline fragment
      if (sel.kind === Kind.INLINE_FRAGMENT) {
        //
      }
    }
    setPath('fields', fields);
  }

  const visitor = visitWithTypeInfo(typeInfo, {
    OperationDefinition(op) {
      normalizedDoc.operation = op.operation;
    },
    VariableDefinition(node) {
      const obj: VariableMeta = {
        name: node.variable.name.value,
      };
      if (node.defaultValue) {
        obj.defaultValue = print(node.defaultValue);
      }
      normalizedDoc.variables.push(obj);
    },
    Field: {
      enter(node) {
        if (node.alias?.value) {
          fieldPath.push(node.alias.value);
          setPath('aliasField', node.name.value);
        } else {
          fieldPath.push(node.name.value);
        }
        let gqlType = typeInfo.getType();
        while (isWrappingType(gqlType)) {
          if (isListType(gqlType)) {
            setPath('list', true);
            fieldPath.push('$idx');
          }
          gqlType = gqlType.ofType;
        }
        if (isAbstractType(gqlType)) {
          setPath(
            'possible',
            schema.getPossibleTypes(gqlType).map((t) => t.name)
          );
        }
        if (isObjectType(gqlType)) {
          const cacheKey = getCacheKey(typePolicies, gqlType);
          if (cacheKey) {
            setPath('cacheKey', cacheKey);
          }
        }
        if (node.arguments?.length) {
          setPath('args', normalizeArgs(node.arguments));
        }
        if (node.selectionSet) {
          addSelectionFields(node.selectionSet);
        }
      },
      leave(node) {
        fieldPath.pop();
        let gqlType = typeInfo.getType();
        while (isWrappingType(gqlType)) {
          if (isListType(gqlType)) {
            fieldPath.pop();
          }
          gqlType = gqlType.ofType;
        }
      },
    },
    InlineFragment: {
      enter(node) {
        if (node.typeCondition?.name) {
          fieldPath.push(`$type_${node.typeCondition?.name.value}`);
          addSelectionFields(node.selectionSet);
        }
      },
      leave(node) {
        if (node.typeCondition?.name.value) {
          fieldPath.pop();
        }
      },
    },
    FragmentDefinition() {
      throw noFragments();
    },
  });

  visit(finalOperation, visitor);

  return normalizedDoc;
}

function noFragments() {
  return new Error('Fragment definitions have been stripped');
}
