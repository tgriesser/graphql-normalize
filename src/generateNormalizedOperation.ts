import { optimizeDocuments } from '@graphql-tools/relay-operation-optimizer';
import { DocumentNode, GraphQLSchema, Kind } from 'graphql';
import { addNormalizingFields } from './addNormalizingFields';

export function generateNormalizedOperation(schema: GraphQLSchema, operation: DocumentNode) {
  let finalOperation = operation;
  if (operation.definitions.some((d) => d.kind === Kind.FRAGMENT_DEFINITION)) {
    finalOperation = optimizeDocuments(schema, [operation], {
      noLocation: true,
    })[0];
  }
  return addNormalizingFields(schema, finalOperation);
}
