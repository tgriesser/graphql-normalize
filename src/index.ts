import { graphqlNormalize } from './graphqlNormalize.js';
export { stringifyVariables } from './stringifyVariables.js';
export { MissingFieldError } from './errors.js';
export type { graphqlNormalizeOptions, NormalizeMetaShape } from './graphqlNormalize.js';
export { graphqlNormalize };
export default graphqlNormalize;
