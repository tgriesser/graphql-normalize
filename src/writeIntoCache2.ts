import type { Immutable } from 'immer';
import type { CacheShape } from './cache';
import type { FieldDef, FieldMeta, NormalizeMetaShape } from './metadataShapes';
import { NO_ARGS, printArgs } from './printArgs';
import { copyList } from './copyList';
import type { FormattedExecutionResult } from 'graphql';
import { makeCacheKey } from './makeCacheKey';

interface WriteQueryIntoCache {
  mutate?: boolean;
  meta: NormalizeMetaShape;
  operationResult: Immutable<FormattedExecutionResult>;
  variableValues: Record<string, any>;
  currentCache?: CacheShape['fields'];
  isEqual?: (a: any, b: any) => boolean;
}

export class MissingFieldError {
  constructor(path: Array<string | number>) {
    throw new Error(`Missing field for: ${path.filter((p) => p !== NO_ARGS)}`);
  }
}

const defaultIsEqual = (a: any, b: any) => {
  return a === b;
};

interface TraverseFieldsConfig {
  into: any;
  existing: any;
  result: any;
  fields: FieldDef[];
  basePath: Array<string | number>;
}

/**
 * Writes a query into the cache
 */
export function writeQueryIntoCache<Q>(config: WriteQueryIntoCache) {
  const {
    meta,
    operationResult,
    variableValues = {},
    currentCache = {} as Q,
    isEqual = defaultIsEqual,
    mutate = true,
  } = config;

  const ensurePath = (obj: any, path: Array<string | number>) => {
    let result = obj;
    for (let i = 0; i < path.length; i++) {
      const key = path[i];
      if (obj[key] === undefined) {
        obj[key] = typeof key === 'number' ? [] : {};
      }
      result = obj[key];
    }
    return result;
  };

  const getIn = (path: Array<string | number>) => {
    for (let i = 0; i < path.length; i++) {
      //
    }
  };

  const setIn = (path: Array<string | number>) => {
    //
  };

  const { fields, operation, variables } = config.meta;

  function shouldSkip(field: FieldMeta) {
    return false;
  }

  interface TraverseFieldsResult {
    additions: Array<string | number>[];
    updates: Array<string | number>[];
    cache: any;
  }

  function traverseFields(config: TraverseFieldsConfig): any {
    const updates: Array<string | number>[] = [];
    const additions: Array<string | number>[] = [];

    const { fields, into = {}, existing, result, basePath } = config;

    const getField = (path: Array<string | number>) => {
      return basePath.concat(path).reduce((acc, key) => acc?.[key], result);
    };

    const getValue = (path: Array<string | number>) => {
      return path.reduce((acc, key) => acc?.[key], existing);
    };

    const setField = (key: string, value: any) => {
      if (into[key] === undefined || !isEqual(into[key], value)) {
        into[key] = value;
        // updates.push();
      }
    };

    const setPath = (path: Array<string | number>, value: any) => {
      let writeInto = into;
      for (let i = 0; i < path.length; i++) {
        const key = path[i];
        const hasNext = i < path.length - 1;
        if (hasNext) {
          if (writeInto[key] === undefined) {
            writeInto[key] = typeof key === 'number' ? [] : {};
            writeInto = writeInto[key];
          } else {
            writeInto = writeInto[key];
          }
        } else if (!isEqual(writeInto[key], value)) {
          if (writeInto[key] === undefined) {
            additions.push(path);
          } else {
            updates.push(path);
          }
          writeInto[key] = value;
        }
      }
    };

    for (const field of fields) {
      if (field === '__typename') {
        setField(field, getField(['__typename']));
      } else if (typeof field === 'string') {
        setPath([field, NO_ARGS], getField([field]));
      } else {
        const argsKey = printArgs(field.args, meta, variableValues);
        const resultName = field.alias ?? field.name;
        const fieldData = getField([resultName]);
        const cacheData = getValue([field.name, argsKey]);

        // console.log({ fieldData, argsKey, resultName });
        // if (shouldSkip(field)) {
        //   if (resultName in into) {
        //     delete into[resultName];
        //     changed = true;
        //     continue;
        //   }
        // }

        // If we have a null value, there's nothing more to do here
        if (fieldData === null) {
          setPath([field.name, argsKey], null);
          continue;
        }

        const handleField = (indexes: number[] = []) => {
          const resultValue = getField([resultName, ...indexes]);
          const cacheKey = field.cacheKey ? makeCacheKey(field.cacheKey, resultValue) : undefined;
          const fieldPath = cacheKey ? [cacheKey] : basePath.concat([resultName]).concat(indexes);

          if (field.fields) {
            const { additions, updates, cache } = traverseFields({
              into: currentCache,
              existing: currentCache,
              fields: field.fields,
              basePath: fieldPath,
              result: result[resultName],
            });
          }

          return cacheKey ? { $ref: cacheKey } : resultValue;
        };

        // If we know we have a list, copy items at the depth of the list
        if (field.list) {
          const { list, modified } = copyList({
            isEqual,
            source: Array.isArray(fieldData) ? fieldData : [],
            depth: field.list,
            target: into[field.name]?.[argsKey] ?? [],
            onValue: (existingValue, newValue, indexes) => {
              return handleField(indexes);
            },
          });
          if (modified) {
            setPath([field.name, argsKey], list);
          }
        } else {
          // ['posts', {first: 10}, { $ref: 1 }]
          setPath([field.name, argsKey], handleField());
        }
      }
    }

    const hasChanges = updates.length || additions.length || mutate;

    return { updates, additions, cache: hasChanges ? into : existing };
  }

  const obj = traverseFields({
    into: mutate ? currentCache : { ...currentCache },
    existing: currentCache,
    fields,
    basePath: [],
    result: operationResult.data,
  });

  return obj;
}
