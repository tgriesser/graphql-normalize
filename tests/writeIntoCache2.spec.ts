import { execute } from 'graphql';
import { describe, expect, it } from 'vitest';
import { generateNormalizedOperation } from '../src/codegen/generateNormalizedOperation';
import { schema } from './fixtures/schema';
import { operation1Doc } from './fixtures/ops';
import { generateNormalizedMetadata } from '../src/codegen/generateNormalizedMetadata';
import { writeQueryIntoCache } from '../src/writeIntoCache2';

describe('writeIntoCache', () => {
  //
  it('builds a cache of values', async () => {
    const meta = generateNormalizedMetadata(schema, operation1Doc);
    const variableValues = {};
    const result = await execute({
      schema,
      variableValues,
      document: generateNormalizedOperation(schema, operation1Doc),
    });
    const { cache, additions, updates } = writeQueryIntoCache({
      meta,
      variableValues,
      operationResult: result,
    });

    expect(cache).toMatchSnapshot();
    expect({ additions, updates }).toMatchSnapshot();
  });
});
