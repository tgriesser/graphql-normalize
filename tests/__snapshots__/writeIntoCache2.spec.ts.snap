// Vitest Snapshot v1

exports[`writeIntoCache > builds a cache of values 1`] = `
{
  "homeItems": {
    "{\\"first\\":\\"10\\"}": {
      "nodes": [
        {
          "__typename": "Comment",
        },
        {
          "__typename": "Blog",
          "id": "QmxvZzox",
          "name": "My Cool #1",
        },
        {
          "__typename": "Post",
          "id": "UG9zdDox",
        },
        {
          "__typename": "User",
        },
        {
          "__typename": "Comment",
        },
        {
          "__typename": "Blog",
          "id": "QmxvZzoy",
          "name": "My Cool #2",
        },
        {
          "__typename": "Post",
          "id": "UG9zdDoy",
        },
        {
          "__typename": "User",
        },
        {
          "__typename": "Comment",
        },
        {
          "__typename": "Blog",
          "id": "QmxvZzoz",
          "name": "My Cool #3",
        },
      ],
    },
  },
  "name": {},
  "nestedString": {
    "$": [
      [
        "a",
        "b",
      ],
      [
        "foo",
      ],
    ],
  },
  "nestedUserList": {
    "$": [
      [
        {
          "$ref": "User:VXNlcjox",
        },
      ],
      [
        {
          "$ref": "User:VXNlcjoy",
        },
      ],
    ],
  },
  "node": {},
  "num": {
    "$": 1,
  },
  "posts": {
    "{\\"first\\":\\"10\\"}": {
      "nodes": [
        {
          "__typename": "Post",
          "id": "UG9zdDox",
          "title": "Some Blog post 1",
        },
        {
          "__typename": "Post",
          "id": "UG9zdDoy",
          "title": "Some Blog post 2",
        },
        {
          "__typename": "Post",
          "id": "UG9zdDoz",
          "title": "Some Blog post 3",
        },
        {
          "__typename": "Post",
          "id": "UG9zdDo0",
          "title": "Some Blog post 4",
        },
        {
          "__typename": "Post",
          "id": "UG9zdDo1",
          "title": "Some Blog post 5",
        },
        {
          "__typename": "Post",
          "id": "UG9zdDo2",
          "title": "Some Blog post 6",
        },
        {
          "__typename": "Post",
          "id": "UG9zdDo3",
          "title": "Some Blog post 7",
        },
        {
          "__typename": "Post",
          "id": "UG9zdDo4",
          "title": "Some Blog post 8",
        },
        {
          "__typename": "Post",
          "id": "UG9zdDo5",
          "title": "Some Blog post 9",
        },
        {
          "__typename": "Post",
          "id": "UG9zdDoxMA==",
          "title": "Some Blog post 10",
        },
      ],
    },
    "{\\"first\\":\\"12\\"}": {
      "edges": [
        {
          "__typename": "QueryPostsConnectionEdge",
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjow",
          "node": {
            "__typename": "Post",
            "author": {
              "__typename": "User",
              "id": "VXNlcjox",
            },
            "id": "UG9zdDox",
          },
        },
        {
          "__typename": "QueryPostsConnectionEdge",
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjox",
          "node": {
            "__typename": "Post",
            "author": {
              "__typename": "User",
              "id": "VXNlcjoy",
            },
            "id": "UG9zdDoy",
          },
        },
        {
          "__typename": "QueryPostsConnectionEdge",
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoy",
          "node": {
            "__typename": "Post",
            "author": {
              "__typename": "User",
              "id": "VXNlcjox",
            },
            "id": "UG9zdDoz",
          },
        },
        {
          "__typename": "QueryPostsConnectionEdge",
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoz",
          "node": {
            "__typename": "Post",
            "author": {
              "__typename": "User",
              "id": "VXNlcjoy",
            },
            "id": "UG9zdDo0",
          },
        },
        {
          "__typename": "QueryPostsConnectionEdge",
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo0",
          "node": {
            "__typename": "Post",
            "author": {
              "__typename": "User",
              "id": "VXNlcjox",
            },
            "id": "UG9zdDo1",
          },
        },
        {
          "__typename": "QueryPostsConnectionEdge",
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo1",
          "node": {
            "__typename": "Post",
            "author": {
              "__typename": "User",
              "id": "VXNlcjoy",
            },
            "id": "UG9zdDo2",
          },
        },
        {
          "__typename": "QueryPostsConnectionEdge",
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo2",
          "node": {
            "__typename": "Post",
            "author": {
              "__typename": "User",
              "id": "VXNlcjox",
            },
            "id": "UG9zdDo3",
          },
        },
        {
          "__typename": "QueryPostsConnectionEdge",
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo3",
          "node": {
            "__typename": "Post",
            "author": {
              "__typename": "User",
              "id": "VXNlcjoy",
            },
            "id": "UG9zdDo4",
          },
        },
        {
          "__typename": "QueryPostsConnectionEdge",
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo4",
          "node": {
            "__typename": "Post",
            "author": {
              "__typename": "User",
              "id": "VXNlcjox",
            },
            "id": "UG9zdDo5",
          },
        },
        {
          "__typename": "QueryPostsConnectionEdge",
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjo5",
          "node": {
            "__typename": "Post",
            "author": {
              "__typename": "User",
              "id": "VXNlcjoy",
            },
            "id": "UG9zdDoxMA==",
          },
        },
        {
          "__typename": "QueryPostsConnectionEdge",
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoxMA==",
          "node": {
            "__typename": "Post",
            "author": {
              "__typename": "User",
              "id": "VXNlcjox",
            },
            "id": "UG9zdDoxMQ==",
          },
        },
        {
          "__typename": "QueryPostsConnectionEdge",
          "cursor": "T2Zmc2V0Q29ubmVjdGlvbjoxMQ==",
          "node": {
            "__typename": "Post",
            "author": {
              "__typename": "User",
              "id": "VXNlcjoy",
            },
            "id": "UG9zdDoxMg==",
          },
        },
      ],
      "nodes": [
        {
          "__typename": "Post",
          "author": {
            "__typename": "User",
            "id": "VXNlcjox",
          },
          "id": "UG9zdDox",
        },
        {
          "__typename": "Post",
          "author": {
            "__typename": "User",
            "id": "VXNlcjoy",
          },
          "id": "UG9zdDoy",
        },
        {
          "__typename": "Post",
          "author": {
            "__typename": "User",
            "id": "VXNlcjox",
          },
          "id": "UG9zdDoz",
        },
        {
          "__typename": "Post",
          "author": {
            "__typename": "User",
            "id": "VXNlcjoy",
          },
          "id": "UG9zdDo0",
        },
        {
          "__typename": "Post",
          "author": {
            "__typename": "User",
            "id": "VXNlcjox",
          },
          "id": "UG9zdDo1",
        },
        {
          "__typename": "Post",
          "author": {
            "__typename": "User",
            "id": "VXNlcjoy",
          },
          "id": "UG9zdDo2",
        },
        {
          "__typename": "Post",
          "author": {
            "__typename": "User",
            "id": "VXNlcjox",
          },
          "id": "UG9zdDo3",
        },
        {
          "__typename": "Post",
          "author": {
            "__typename": "User",
            "id": "VXNlcjoy",
          },
          "id": "UG9zdDo4",
        },
        {
          "__typename": "Post",
          "author": {
            "__typename": "User",
            "id": "VXNlcjox",
          },
          "id": "UG9zdDo5",
        },
        {
          "__typename": "Post",
          "author": {
            "__typename": "User",
            "id": "VXNlcjoy",
          },
          "id": "UG9zdDoxMA==",
        },
        {
          "__typename": "Post",
          "author": {
            "__typename": "User",
            "id": "VXNlcjox",
          },
          "id": "UG9zdDoxMQ==",
        },
        {
          "__typename": "Post",
          "author": {
            "__typename": "User",
            "id": "VXNlcjoy",
          },
          "id": "UG9zdDoxMg==",
        },
      ],
    },
  },
}
`;

exports[`writeIntoCache > builds a cache of values 2`] = `
{
  "additions": [
    [
      "posts",
      "{\\"first\\":\\"12\\"}",
    ],
    [
      "posts",
      "{\\"first\\":\\"10\\"}",
    ],
    [
      "homeItems",
      "{\\"first\\":\\"10\\"}",
    ],
    [
      "num",
      "$",
    ],
    [
      "nestedString",
      "$",
    ],
    [
      "nestedUserList",
      "$",
    ],
  ],
  "updates": [],
}
`;
